src/python/grpcio_tests/tests/testing/_time_test.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2017 gRPC authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import threading
+import time
+import random
+import unittest
+
+import grpc_testing
+
+_QUANTUM = 0.3
+_MANY = 10000
+
+
+class TimeTest(object):
+
+    def test_sleep_for(self):
+        start_time = self._time.time()
+        self._time.sleep_for(_QUANTUM)
+        end_time = self._time.time()
+
+        self.assertLessEqual(start_time + _QUANTUM, end_time)
+
+    def test_sleep_until(self):
+        start_time = self._time.time()
+        self._time.sleep_until(start_time + _QUANTUM)
+        end_time = self._time.time()
+
+        self.assertLessEqual(start_time + _QUANTUM, end_time)
+
+    def test_call_in(self):
+        event = threading.Event()
+
+        self._time.call_in(event.set, _QUANTUM)
+        self._time.sleep_for(_QUANTUM * 2)
+
+        self.assertTrue(event.is_set())
+
+    def test_call_at(self):
+        event = threading.Event()
+
+        self._time.call_at(event.set, self._time.time() + _QUANTUM)
+        self._time.sleep_for(_QUANTUM * 2)
+
+        self.assertTrue(event.is_set())
+
+    def test_cancel(self):
+        event = threading.Event()
+
+        future = self._time.call_in(event.set, _QUANTUM * 2)
+        self._time.sleep_for(_QUANTUM)
+        cancelled = future.cancel()
+        self._time.sleep_for(_QUANTUM * 2)
+
+        if cancelled:
```

Comment:
I don't think you can escape the raciness. The test passes either way, right?
I think you should just comment why cancelled might be false.